serde-support/src/main/java/io/micronaut/serde/support/deserializers/DeserBeanSubtypeInfo.java

```diff
@@ -0,0 +1,153 @@
+import io.micronaut.serde.support.util.SubtypeInfo;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * The subtype info.
+ *
+ * @param beanType    The beanType
+ * @param subtypes    The subtypes
+ * @param info        The subtype info
+ * @param defaultType The default type
+ * @param <T>         The bean type
+ * @author Denis Stepanov
+ */
+@Internal
+record DeserBeanSubtypeInfo<T>(
+    Class<T> beanType,
+    @NonNull
+    Map<String, DeserBean<? extends T>> subtypes,
+    SubtypeInfo info,
+    @Nullable
+    DeserBean<? extends T> defaultType
+) {
+
+    /**
+     * Find the {@link DeserBean} for discriminator of not provided one.
+     *
+     * @param discriminatorValue The discriminator value
+     * @return The {@link DeserBean}
+     */
+    @NonNull
+    DeserBean<? extends T> findDeserBean(@Nullable String discriminatorValue) throws SerdeException {
+        DeserBean<? extends T> deserBean;
+        if (discriminatorValue == null) {
+            deserBean = defaultType;
+        } else {
+            deserBean = subtypes.getOrDefault(discriminatorValue, defaultType);
+        }
+        if (deserBean == null) {
+            throw unknownSuperTypeException();
+        }
+        return deserBean;
+    }
+
+    /**
+     * @return Creates unknown supertype exception
+     */
+    @NonNull
+    public SerdeException unknownSuperTypeException() {
+        return new SerdeException("Could not resolve subtype of ["
+            + beanType.getName() + "] missing type id property '" + info.discriminatorName() + "'");
+    }
+
+    @Nullable
+    static <T> DeserBeanSubtypeInfo<T> create(@Nullable SubtypeInfo subtypeInfo,
+                                              BeanIntrospection<T> introspection,
+                                              Deserializer.DecoderContext decoderContext,
+                                              DeserBean<T> superTypeDeserBean,
+                                              DeserializationConfiguration deserializationConfiguration,
+                                              DeserBeanRegistry deserBeanRegistry) throws SerdeException {
+
+        if (subtypeInfo == null) {
+            return null;
+        }
+
+        final Class<T> superType = introspection.getBeanType();
+        final Collection<BeanIntrospection<? extends T>> subtypeIntrospections =
+            decoderContext.getDeserializableSubtypes(superType);
+        Map<String, DeserBean<? extends T>> subtypes = CollectionUtils.newHashMap(subtypeIntrospections.size());
+        Class<? extends T> defaultType = introspection.classValue(DefaultImplementation.class)
+            .orElseGet(() -> introspection.classValue(SerdeConfig.SerSubtyped.DEFAULT_IMPL).orElse(null));
+        if (defaultType == null && !deserializationConfiguration.isAvoidSupertypeSubtype()) {
+            // Jackson always requires to use `defaultImpl` but our initial implementation always takes the supertype
```

Comment:
should there be some kind of failure if defaultImpl is not specified?

Reply:
It will fail at runtime when the subtypes fails to be resolved.